### PR TITLE
ARROW-13242: [C++] Improve random generation of decimal arrays

### PR DESCRIPTION
**Pull request by pitrou**
- Allow precisions larger than a single uint64
- Implement decimal256 generation
- Add validity tests

### CODE REVIEW

**Comment by pitrou** [comment]
@mathyingzhou This may interest you.

**Comment by pitrou** [comment]
Hmm, looks like the algorithm needs fixing.

File: cpp/src/arrow/testing/random.cc [hunk 1]
```diff
@@ -229,34 +230,95 @@ std::shared_ptr<Array> RandomArrayGenerator::Float64(int64_t size, double min, d
...
+
+    const DecimalValue kDigitsMultiplier =
+        DecimalValue::GetScaleMultiplier(kMaxDigitsInInteger);
+
+    for (int64_t i = 0; i < size; ++i) {
+      if (sign_array->IsValid(i)) {
+        DecimalValue dec_value{0};
+        for (int j = 0; j < kNumIntegers; ++j) {
+          dec_value =
+              dec_value * kDigitsMultiplier + DecimalValue(digit_arrays[j]->Value(i));
```
Comment (cyb70289) [inline cpp/src/arrow/testing/random.cc:291]: So this is endianness agnostic. And the result will always be positive as 10^19 has 63.12 bits, ceil(63.12*2) < 128.

**Review by cyb70289 — APPROVED** [review]
LGTM, +1

File: cpp/src/arrow/testing/random.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/testing/random.cc:291]: `< 127` even ;-)

**Comment by pitrou** [comment]
Rebased.

**Pull request closed without merging**